### PR TITLE
Add `FriendlyExceptionInterface::class`, add more tests.

### PR DESCRIPTION
**Pull request by terabytesoftw**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

### CODE REVIEW

File: src/Exception/AttributeNotSetException.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    }
+
+    public function getSolution(): ?string
+    {
+        return <<<SOLUTION
+            You can configure the `attribute` in two ways, the first way is through the widgets using the `for()`
+        method `Text::widget()->for(FormModel::class, attribute)`, where the second argument is the `attribute`,
+        the second way is through the `Field::class` where in each field type the second argument is the
+        `attribute`, `Field::widget->text(FormModel::class, attribute)`.
+        SOLUTION;
```
Comment (samdark) [inline src/Exception/AttributeNotSetException.php:24]:
```suggestion
            You can configure the `attribute` in two ways. The first way is through the widgets using the `for()`
        method `Text::widget()->for(FormModel::class, attribute)`, where the second argument is the `attribute`.
        The second way is through the `Field::class` where in each field type the second argument is the
        `attribute`, `Field::widget->text(FormModel::class, attribute)`.
        SOLUTION;
```

File: src/Exception/FormModelNotSetException.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        return 'Failed to create widget because form model is not set.';
+    }
+
+    public function getSolution(): ?string
+    {
+        return <<<SOLUTION
+            You can configure the `FormModel::class` in two ways, the first way is through the widgets using the `for()`
+        method `Text::widget()->for(FormModel::class, attribute)`, where the first argument is the `FormModel::class`,
+        the second way is through the `Field::class` where in each field type the first argument is the
+        `FormModel::class`, `Field::widget->text(FormModel::class, attribute)`.
```
Comment (samdark) [inline src/Exception/FormModelNotSetException.php:23]:
```suggestion
            You can configure the `FormModel::class` in two ways. The first way is through the widgets using the `for()`
        method `Text::widget()->for(FormModel::class, attribute)`, where the first argument is the `FormModel::class`.
        The second way is through the `Field::class` where in each field type the first argument is the
        `FormModel::class`, `Field::widget->text(FormModel::class, attribute)`.
```

File: src/Widget/FieldPart/Hint.php [hunk 3]
```diff
@@ -138,7 +140,7 @@ protected function run(): string
     private function getAttribute(): string
     {
         if ($this->attribute === '') {
-            throw new InvalidArgumentException('Attribute is not set.');
+            throw new AttributeNotSetException('Failed to create widget because attribute is not set.');
```
Comment (samdark) [inline src/Widget/FieldPart/Hint.php:143]: Are these the same messages as in the class itself?

File: src/Widget/FieldPart/Hint.php [hunk 4]
```diff
@@ -147,7 +149,7 @@ private function getAttribute(): string
     private function getFormModel(): FormModelInterface
     {
         if ($this->formModel === null) {
-            throw new InvalidArgumentException('Form model is not set.');
+            throw new FormModelNotSetException('Failed to create widget because form model is not set.');
```
Comment (samdark) [inline src/Widget/FieldPart/Hint.php:152]: Are these the same messages as in the class itself? How about removing constructor argument and making these part of the exception class?

**Review by samdark — CHANGES_REQUESTED** [review]

File: src/Exception/AttributeNotSetException.php [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+namespace Yiisoft\Form\Exception;
+
+use InvalidArgumentException;
+use Yiisoft\FriendlyException\FriendlyExceptionInterface;
+
+final class AttributeNotSetException extends InvalidArgumentException implements FriendlyExceptionInterface
+{
+    public function getName(): string
+    {
+        return 'Failed to create widget because attribute is not set.';
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Exception/AttributeNotSetException.php:14]: done.

File: src/Exception/AttributeNotSetException.php (hunk 1 above)
Comment (terabytesoftw) [reply src/Exception/AttributeNotSetException.php:24]: done.

File: src/Exception/FormModelNotSetException.php (hunk 2 above)
Comment (terabytesoftw) [reply src/Exception/FormModelNotSetException.php:23]: done.

File: src/Widget/FieldPart/Hint.php (hunk 4 above)
Comment (terabytesoftw) [reply src/Widget/FieldPart/Hint.php:152]: done.

Comment (terabytesoftw) [reply src/Widget/FieldPart/Hint.php:152]: done.

**Pull request merged**